### PR TITLE
[DRAFT] refactoring experiments

### PR DESCRIPTION
**Pull request by cmaglie**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [ ] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

## What kind of change does this PR introduce?

I want to run integration tests to check the feasibility of a quite big refactoring.

### CODE REVIEW

File: commands/service_settings_test.go
Comment (alessio-perugini) [inline commands/service_settings_test.go:1]: We should also test that envvars takes precedence, over some values in the config.

File: commands/service_monitor.go
Comment (alessio-perugini) [inline commands/service_monitor.go:1]: IIRC this part should be covered by the integration tests that uses the mocked monitor right?

**Review by alessio-perugini — CHANGES_REQUESTED** [review]
Great work I like the direction we're going to.
I'll level below 2 things here that we should decide:
- We want to support 2 `arduino-cli` config file (json, and yaml)?
- In the protobuf responses some time we use oneof other time we don't. Might be worth to review all the Respones and make them behave similar, before we release the 1.0.0

File: rpc/internal/client_example/main.go [hunk 1]
```diff
@@ -245,84 +244,84 @@ func callVersion(client rpc.ArduinoCoreServiceClient) {
...
+	// 		Key:      "directories",
+	// 		JsonData: `{"data": "` + dataDir + `", "downloads": "` + path.Join(dataDir, "staging") + `", "user": "` + path.Join(dataDir, "sketchbook") + `"}`,
+	// 	})
 
-	if err != nil {
-		log.Fatalf("Error setting settings value: %s", err)
-	}
+	// if err != nil {
+	// 	log.Fatalf("Error setting settings value: %s", err)
+	// }
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply rpc/internal/client_example/main.go:255]: Oh right, let's wait a bit for the review, I'll rewrite it as the last step before merging.

File: internal/cli/config/set.go [hunk 2]
```diff
@@ -37,40 +38,40 @@ func initSetCommand() *cobra.Command {
...
 
-	configuration.Settings.Set(key, value)
-
-	if err := configuration.Settings.WriteConfig(); err != nil {
-		feedback.Fatal(tr("Writing config file: %v", err), feedback.ErrGeneric)
+	if _, err := srv.SettingsSetValue(ctx, req); err != nil {
+		feedback.Fatal(tr("Error setting value: %v", err), feedback.ErrGeneric)
 	}
+
+	saveConfiguration(ctx, srv)
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply internal/cli/config/set.go:76]: Viper supports a lot of formats: https://arduino.github.io/arduino-cli/0.35/configuration/#supported-formats

BTW I don't see why we should support all of this, it's confusing and chaotic. I'm for dropping the support for all formats and only keeping YAML.

File: rpc/cc/arduino/cli/commands/v1/settings.proto [hunk 3]
```diff
@@ -19,51 +19,145 @@ package cc.arduino.cli.commands.v1;
 
 option go_package = "github.com/arduino/arduino-cli/rpc/cc/arduino/cli/commands/v1;commands";
 
-message SettingsGetAllResponse {
-  // The settings, in JSON format.
-  string json_data = 1;
+// Configuration to apply to the given instance.
+// Any missing field will be kept at the default value.
+message Configuration {
```
[the remark this answers is not in the document]
Comment (alessio-perugini) [reply rpc/cc/arduino/cli/commands/v1/settings.proto:24]: > while loading a configuration the CLI now fails if the config contains "unknown" keys, instead it should just ignore them.

Agreed

File: internal/cli/config/set.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmaglie) [reply internal/cli/config/set.go:76]: In the end, we will drop support for config file formats different from YAML, I'll update the PR accordingly.

File: internal/cli/config/config.go [hunk 4]
```diff
@@ -16,45 +16,80 @@
...
+}
+
+// SetConfigFile sets the configuration file path in the context
+func SetConfigFile(ctx context.Context, configFile string) context.Context {
+	return context.WithValue(ctx, ctxValue("config_file"), configFile)
+}
+
+func saveConfiguration(ctx context.Context, srv rpc.ArduinoCoreServiceServer) {
+	var outConfig []byte
+	if res, err := srv.ConfigurationSave(ctx, &rpc.ConfigurationSaveRequest{SettingsFormat: "yaml"}); err != nil {
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply internal/cli/config/config.go:85]: In the end, we will drop support for config file formats different from YAML, I'll update the PR accordingly.

File: commands/service_upload.go [hunk 5]
```diff
@@ -154,6 +176,20 @@ func Upload(ctx context.Context, req *rpc.UploadRequest, outStream io.Writer, er
...
+	})
+	defer outStream.Close()
+	errStream := feedStreamTo(func(data []byte) {
+		syncSend.Send(&rpc.UploadResponse{
+			Message: &rpc.UploadResponse_ErrStream{ErrStream: data},
+		})
+	})
+	defer errStream.Close()
+	// TODO: inject context
+	// ctx := stream.Context()
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/service_upload.go:192]: Yep, there is another ton of those missing :-/... on it..

File: commands/service_settings.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+	case "json":
+		err := json.Unmarshal([]byte(req.GetEncodedSettings()), s.settings)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling settings: %v", err)
+		}
+	default:
+		return nil, &cmderrors.InvalidArgumentError{Message: fmt.Sprintf("unsupported format: %s", req.GetSettingsFormat())}
+	}
+
+	configuration.BindEnvVars(s.settings)
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/service_settings.go:205]: > I remember that env vars takes precedence over values written inside the config.

Correct, for that reason the binding is done after the unmarshal of the config file.

Maybe the method should be called `SetFromEnvVars` or `UnmarshalFromEnvVars`...

File: commands/service_monitor.go
Comment (cmaglie) [reply commands/service_monitor.go:1]: There are some tests for edge cases on monitor close signaling, and for monitor configuration change (baudrate, etc) but not for stress testing the data streams.

File: commands/service_settings_test.go
Comment (cmaglie) [reply commands/service_settings_test.go:1]: There are some integration tests already 🤖 💯

File: main.go [hunk 7]
```diff
@@ -16,19 +16,54 @@
...
-	if err := arduinoCmd.Execute(); err != nil {
+	// Create a new ArduinoCoreServer
+	srv := commands.NewArduinoCoreServer()
+
+	// Search for the configuration file in the command line arguments and in the environment
+	configFile := configuration.FindConfigFileInArgsFallbackOnEnv(os.Args)
+	ctx := config.SetConfigFile(context.Background(), configFile)
+
+	// Read the settings from the configuration file
+	openReq := &rpc.ConfigurationOpenRequest{SettingsFormat: "yaml"}
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply main.go:42]: We are removing this possibility, I've updated the PR.

File: internal/cli/config/init.go [hunk 8]
```diff
@@ -71,46 +75,95 @@ func runInitCommand(cmd *cobra.Command, args []string) {
...
-	configuration.SetDefaults(newSettings)
-	configuration.BindFlags(cmd, newSettings)
+	tmpSrv := commands.NewArduinoCoreServer()
 
-	for _, url := range newSettings.GetStringSlice("board_manager.additional_urls") {
-		if strings.Contains(url, ",") {
-			feedback.Fatal(tr("Urls cannot contain commas. Separate multiple urls exported as env var with a space:\n%s", url),
-				feedback.ErrGeneric)
-		}
+	if _, err := tmpSrv.ConfigurationOpen(ctx, &rpc.ConfigurationOpenRequest{SettingsFormat: "yaml", EncodedSettings: ""}); err != nil {
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply internal/cli/config/init.go:104]: We are removing this possibility, I've updated the PR.

File: rpc/cc/arduino/cli/commands/v1/settings.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmaglie) [reply rpc/cc/arduino/cli/commands/v1/settings.proto:24]: I changed the config loader to ignore config "errors" and produce warnings instead.

File: commands/service_platform_upgrade.go [hunk 9]
```diff
@@ -13,20 +13,45 @@
...
+			if r.GetProgress() != nil {
+				downloadCB(r.GetProgress())
+			}
+			if r.GetTaskProgress() != nil {
+				taskCB(r.GetTaskProgress())
+			}
+			if r.GetPlatform() != nil {
+				resp = r.GetPlatform()
+			}
+			return nil
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/service_platform_upgrade.go:42]: You're right, I'm going to move this into a different issue.

File: internal/go-configmap/configuration.go
[the remark this answers is not in the document]
Comment (cmaglie) [reply internal/go-configmap/configuration.go:1]: I've added missing tests and also improved conversion!

File: internal/go-configmap/types.go
[the remark this answers is not in the document]
Comment (cmaglie) [reply internal/go-configmap/types.go:1]: I've added missing test and improved conversion.

File: commands/instances.go [hunk 10]
```diff
@@ -394,66 +421,118 @@ func Init(req *rpc.InitRequest, responseCallback func(r *rpc.InitResponse)) erro
...
-	if err := globals.LibrariesIndexResource.Download(indexDir, downloadCB); err != nil {
-		return nil, err
+	// TODO: pass context
+	// ctx := stream.Context()
+	config, err := s.settings.DownloaderConfig()
+	if err != nil {
+		return err
+	}
+	if err := globals.LibrariesIndexResource.Download(indexDir, downloadCB, config); err != nil {
+		resultCB(rpc.IndexUpdateReport_STATUS_FAILED)
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/instances.go]: Yes, but this requires a small refactor of the Downloader library.
I'm moving this into a separate issue...

File: commands/service_platform_download.go [hunk 11]
```diff
@@ -48,18 +60,23 @@ func PlatformDownload(ctx context.Context, req *rpc.PlatformDownloadRequest, dow
...
 	}
 
-	if err := pme.DownloadPlatformRelease(platform, nil, downloadCB); err != nil {
-		return nil, err
+	downloadCB := func(p *rpc.DownloadProgress) { syncSend.Send(&rpc.PlatformDownloadResponse{Progress: p}) }
+
+	// TODO: pass context
+	// ctx := stream.Context()
+	if err := pme.DownloadPlatformRelease(platform, downloadCB); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/service_platform_download.go:71]: Yes, but this requires a small refactor of the Downloader library.
I'm moving this into a separate issue...

File: commands/service_board_list.go [hunk 12]
```diff
@@ -83,10 +84,7 @@ func apiByVidPid(vid, pid string) ([]*rpc.BoardListItem, error) {
 	req, _ := http.NewRequest("GET", url, nil)
 	req.Header.Set("Content-Type", "application/json")
 
-	// TODO: use proxy if set
-
-	httpClient, err := httpclient.New()
-
+	httpClient, err := settings.NewHttpClient()
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/service_board_list.go:87]: Yep!

File: commands/service_settings.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmaglie) [reply commands/service_settings.go]: I've renamed it to `InjectEnvVars` to keep consistency with `go-configmap`

**Review by alessio-perugini — APPROVED** [review]
LGTM

I'd say after the PR that unifies the usage of `oneOf` in our API, we can schedule a round of manual testing to catch any regression (if any) not seen by our integration tests.

File: internal/integrationtest/config/config_test.go [hunk 13]
```diff
@@ -893,4 +893,12 @@ build.unk: 123
 	out, _, err = cli.Run("config", "get", "locale", "--config-file", invalidConfig.String())
 	require.NoError(t, err)
 	require.Equal(t, "en", strings.TrimSpace(string(out)))
+
+	// Run "config get" with a configuration contatinng a null array
```
Comment (umbynos) [inline internal/integrationtest/config/config_test.go:897]: typo

**Review by umbynos — APPROVED** [review]
Overall LGTM, I did not find any major problem.
Should we update also this: https://github.com/arduino/arduino-cli/blob/master/docs/integration-options.md ?

**Comment by cmaglie** [comment]
> Should we update also this: https://github.com/arduino/arduino-cli/blob/master/docs/integration-options.md?

Yes, BTW I'm doing it in another branch as part of a general revamp of the documentation.

File: .github/workflows/check-go-task.yml [hunk 14]
```diff
@@ -237,7 +237,7 @@ jobs:
         module:
           - path: internal/arduino/discovery/discovery_client
           - path: rpc/internal/client_example
-          - path: commands/daemon/term_example
```
Comment (EMMANUEL2811UNITEC) [inline .github/workflows/check-go-task.yml:240]: 240c8b3e5a87aafc6255afdc943326e3ca1a8dc3

**Review by EMMANUEL2811UNITEC — APPROVED** [review]
